### ISSUE TITLE
Pytorch can't be installed on Python3.8+

### ISSUE THREAD

**Issue by surak** [pytorch/pytorch#47849]
## To Reproduce

1. Install PyTorch from source. It gives the errors:

```
torch 1.7.0 requires dataclasses, which is not installed.
torch 1.7.0 requires typing-extensions, which is not installed.
```
Try to install dataclasses, and it comes with this:
2.ERROR: Package 'dataclasses' requires a different Python: 3.8.5 not in '>=3.6, <3.7'

## Expected behavior

Remove dependencies on obsolete packages with current versions of python

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.0
 - OS (e.g., Linux): Centos8
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Python version:3.8.5
 - CUDA/cuDNN version: 11.0, 8.0.2.39
 - GPU models and configuration: 3700 NVIDIA a100 gpus (JUWELS BOOSTER Supercomputer)

Setup.py should check python version for such deps

cc @ezyang @seemethere @malfet @walterddr

**Comment by skyline75489** [comment]
There is actually a issue for this #46930.

Are you using a non-official pypi ? Most people won’t hit the “ dataclasses' requires a different Python” when using official pypi.

**Comment by seemethere** [comment]
Yes, how are you actually installing this?

**Comment by wjjwsl** [comment]
Hi I have the same problem and I can reproduce it. I used MobaXterm on windows 64 bits. Because of the privacy setting of the computer, I cannot install packages from website, and I have to download thee .whl file and install it locally. I am using python3.8, and when I install dataclasses by `pip install dataclasses-0.8-py3-none-any.whl`, I got the same error.

Also, because of the private setting, I cannot run these commands to solve this problem:
`pip install -r https://raw.githubusercontent.com/pytorch/pytorch/master/requirements.txt
pip install <torch_wheel> --no-deps`

Could anyone give me some suggestion about this? Thank you!

**Comment by ezyang** [comment]
dataclasses comes with python 3.8, you don't need to install it. What happens if you skip trying to do that step and install anyway?

**Comment by rgommers** [comment]
The runtime dependencies declared in `setup.py` are:
```python
 install_requires = [
     'typing_extensions',
     'dataclasses; python_version < "3.7"'
 ]
```
Those are correct. What could be a cause if you're still running into the `dataclasses` issue is that you have a really old `setuptools`, you need `>= 36.2` for it to recognize the `; python_version < "3.7"` part of the requirement.

I have also downloaded released wheels and checked the `METADATA`, that shows that 1.7.0 does miss the correct Python version specifier, it contains:
```
Requires-Python: >=3.6.1
Description-Content-Type: text/markdown
Requires-Dist: future
Requires-Dist: typing-extensions
Requires-Dist: dataclasses
Requires-Dist: numpy
```

For 1.7.1 this was fixed:
```
Requires-Python: >=3.6.2
Description-Content-Type: text/markdown
Requires-Dist: typing-extensions
Requires-Dist: numpy
Requires-Dist: dataclasses ; python_version < "3.7"
```
And for completeness I also checked for 1.8.1, still has the same metadata as 1.7.1.

tl;dr the bug report was valid for `1.7.0`. and the issue was fixed in `1.7.1`. So I'll close this issue.
